File: scripts/generate_trend_chart.py
#!/usr/bin/env python3
"""
生成性能趋势图
"""

import json
import matplotlib.pyplot as plt
import matplotlib.dates as mdates
from datetime import datetime
import os
import sys

def generate_trend_chart(data_files, output_file):
    """
    生成性能趋势图
    
    Args:
        data_files: 历史数据文件列表
        output_file: 输出文件路径
    """
    # 加载历史数据
    historical_data = []
    for data_file in data_files:
        try:
            with open(data_file, 'r') as f:
                historical_data.append(json.load(f))
        except Exception as e:
            print(f"Warning: Failed to load {data_file}: {e}")
    
    if not historical_data:
        print("No valid data files found")
        return
    
    # 按日期排序
    historical_data.sort(key=lambda x: x['timestamp'])
    
    # 提取数据
    dates = [datetime.strptime(d['timestamp'], '%Y-%m-%dT%H:%M:%SZ') for d in historical_data]
    
    # 创建图表
    fig, axes = plt.subplots(2, 2, figsize=(16, 12))
    fig.suptitle('UVHTTP Performance Trends', fontsize=16, fontweight='bold')
    
    # 1. 吞吐量趋势
    ax1 = axes[0, 0]
    rps_low = []
    rps_medium = []
    rps_high = []
    
    for d in historical_data:
        for s in d.get('test_scenarios', []):
            if s['name'] == 'low_concurrent':
                rps_low.append(s['results']['rps']['value'])
            elif s['name'] == 'medium_concurrent':
                rps_medium.append(s['results']['rps']['value'])
            elif s['name'] == 'high_concurrent':
                rps_high.append(s['results']['rps']['value'])
    
    if rps_low:
        ax1.plot(dates, rps_low, marker='o', label='Low Concurrent (10)', linewidth=2)
    if rps_medium:
        ax1.plot(dates, rps_medium, marker='s', label='Medium Concurrent (50)', linewidth=2)
    if rps_high:
        ax1.plot(dates, rps_high, marker='^', label='High Concurrent (100)', linewidth=2)
    
    ax1.set_xlabel('Date', fontsize=12)
    ax1.set_ylabel('Requests Per Second', fontsize=12)
    ax1.set_title('Throughput Trends', fontsize=14, fontweight='bold')
    ax1.legend()
    ax1.grid(True, alpha=0.3)
    ax1.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m-%d'))
    plt.setp(ax1.xaxis.get_majorticklabels(), rotation=45, ha='right')
    
    # 2. 延迟趋势（简化版，只显示平均延迟）
    ax2 = axes[0, 1]
    latency_low = []
    latency_medium = []
    latency_high = []
    
    for d in historical_data:
        for s in d.get('test_scenarios', []):
            if s['name'] == 'low_concurrent':
                latency_low.append(float(s['results']['latency_avg']['value'].replace('us', '').replace('ms', '')))
            elif s['name'] == 'medium_concurrent':
                latency_medium.append(float(s['results']['latency_avg']['value'].replace('us', '').replace('ms', '')))
            elif s['name'] == 'high_concurrent':
                latency_high.append(float(s['results']['latency_avg']['value'].replace('us', '').replace('ms', '')))
    
    if latency_low:
        ax2.plot(dates, latency_low, marker='o', label='Low Concurrent', linewidth=2)
    if latency_medium:
        ax2.plot(dates, latency_medium, marker='s', label='Medium Concurrent', linewidth=2)
    if latency_high:
        ax2.plot(dates, latency_high, marker='^', label='High Concurrent', linewidth=2)
    
    ax2.set_xlabel('Date', fontsize=12)
    ax2.set_ylabel('Average Latency', fontsize=12)
    ax2.set_title('Latency Trends', fontsize=14, fontweight='bold')
    ax2.legend()
    ax2.grid(True, alpha=0.3)
    ax2.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m-%d'))
    plt.setp(ax2.xaxis.get_majorticklabels(), rotation=45, ha='right')
    
    # 3. P99 延迟趋势
    ax3 = axes[1, 0]
    p99_low = []
    p99_medium = []
    p99_high = []
    
    for d in historical_data:
        for s in d.get('test_scenarios', []):
            if s['name'] == 'low_concurrent':
                p99_value = s['results'].get('latency_p99', 'N/A')
                if p99_value != 'N/A':
                    p99_low.append(float(p99_value.replace('us', '').replace('ms', '')))
            elif s['name'] == 'medium_concurrent':
                p99_value = s['results'].get('latency_p99', 'N/A')
                if p99_value != 'N/A':
                    p99_medium.append(float(p99_value.replace('us', '').replace('ms', '')))
            elif s['name'] == 'high_concurrent':
                p99_value = s['results'].get('latency_p99', 'N/A')
                if p99_value != 'N/A':
                    p99_high.append(float(p99_value.replace('us', '').replace('ms', '')))
    
    if p99_low:
        ax3.plot(dates, p99_low, marker='o', label='Low Concurrent', linewidth=2)
    if p99_medium:
        ax3.plot(dates, p99_medium, marker='s', label='Medium Concurrent', linewidth=2)
    if p99_high:
        ax3.plot(dates, p99_high, marker='^', label='High Concurrent', linewidth=2)
    
    ax3.set_xlabel('Date', fontsize=12)
    ax3.set_ylabel('P99 Latency', fontsize=12)
    ax3.set_title('P99 Latency Trends', fontsize=14, fontweight='bold')
    ax3.legend()
    ax3.grid(True, alpha=0.3)
    ax3.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m-%d'))
    plt.setp(ax3.xaxis.get_majorticklabels(), rotation=45, ha='right')
    
    # 4. 测试场景数量趋势
    ax4 = axes[1, 1]
    total_scenarios = [len(d.get('test_scenarios', [])) for d in historical_data]
    
    ax4.plot(dates, total_scenarios, marker='o', color='blue', label='Total Scenarios', linewidth=2)
    ax4.set_xlabel('Date', fontsize=12)
    ax4.set_ylabel('Number of Scenarios', fontsize=12)
    ax4.set_title('Test Scenarios Trends', fontsize=14, fontweight='bold')
    ax4.legend()
    ax4.grid(True, alpha=0.3)
    ax4.xaxis.set_major_formatter(mdates.DateFormatter('%Y-%m-%d'))
    plt.setp(ax4.xaxis.get_majorticklabels(), rotation=45, ha='right')
    
    plt.tight_layout()
    plt.savefig(output_file, dpi=300, bbox_inches='tight')
    print(f"Performance trend chart saved to {output_file}")

if __name__ == '__main__':
    if len(sys.argv) < 3:
        print("Usage: python3 generate_trend_chart.py <data_files> <output_file>")
        sys.exit(1)
    
    data_files = sys.argv[1:-1]
    output_file = sys.argv[-1]
    generate_trend_chart(data_files, output_file)

File: scripts/generate_trend_chart.js
#!/usr/bin/env node
/**
 * Generate performance trend chart using Mermaid
 * 
 * This script generates Mermaid charts for performance trends
 * including throughput, latency, and P99 latency trends
 */

const fs = require('fs').promises;

/**
 * Load JSON file
 * @param {string} filepath - Path to JSON file
 * @returns {Promise<Object>} Parsed JSON object
 */
async function loadJsonFile(filepath) {
  try {
    const content = await fs.readFile(filepath, 'utf-8');
    return JSON.parse(content);
  } catch (error) {
    console.warn(`Warning: Failed to load ${filepath}: ${error.message}`);
    return null;
  }
}

/**
 * Parse date string to timestamp
 * @param {string} dateString - Date string in ISO format
 * @returns {number} Unix timestamp
 */
function parseDate(dateString) {
  const date = new Date(dateString);
  return date.getTime();
}

/**
 * Format date for display
 * @param {string} dateString - Date string in ISO format
 * @returns {string} Formatted date string
 */
function formatDate(dateString) {
  const date = new Date(dateString);
  return date.toISOString().split('T')[0]; // YYYY-MM-DD
}

/**
 * Extract scenario data from historical data
 * @param {Array} historicalData - Historical performance data
 * @param {string} scenarioName - Scenario name
 * @param {string} metric - Metric name (rps, latency_avg, latency_p99)
 * @returns {Array} Array of { date, value }
 */
function extractScenarioData(historicalData, scenarioName, metric) {
  const data = [];
  for (const d of historicalData) {
    const date = formatDate(d.timestamp || d.date || d.updated_at);
    for (const s of d.test_scenarios || []) {
      if (s.name === scenarioName) {
        let value = null;
        if (metric === 'rps') {
          value = s.results?.rps?.value;
        } else if (metric === 'latency_avg') {
          value = s.results?.latency_avg?.value;
        } else if (metric === 'latency_p99') {
          value = s.results?.latency_p99?.value;
        }
        
        if (value !== null && value !== undefined && value !== 'N/A') {
          // Remove unit suffixes (us, ms)
          if (typeof value === 'string') {
            value = parseFloat(value.replace('us', '').replace('ms', ''));
          }
          data.push({ date, value });
        }
      }
    }
  }
  return data;
}

/**
 * Generate Mermaid xychart for throughput trends
 * @param {Array} historicalData - Historical performance data
 * @returns {string} Mermaid xychart code
 */
function generateThroughputChart(historicalData) {
  const rpsLow = extractScenarioData(historicalData, 'low_concurrent', 'rps');
  const rpsMedium = extractScenarioData(historicalData, 'medium_concurrent', 'rps');
  const rpsHigh = extractScenarioData(historicalData, 'high_concurrent', 'rps');
  
  // Find min and max values for y-axis
  const allValues = [
    ...rpsLow.map(d => d.value),
    ...rpsMedium.map(d => d.value),
    ...rpsHigh.map(d => d.value)
  ];
  const minValue = Math.min(...allValues) * 0.9;
  const maxValue = Math.max(...allValues) * 1.1;
  
  let chart = '```mermaid\nxychart-beta\n    title "Throughput Trends"\n';
  chart += `    x-axis [${rpsLow.map(d => d.date).join(', ')}]\n`;
  chart += `    y-axis "${minValue}" --> "${maxValue}"\n`;
  chart += '    line [';
  
  if (rpsLow.length > 0) {
    chart += rpsLow.map(d => d.value).join(', ');
  } else {
    chart += '0';
  }
  chart += ']\n';
  
  if (rpsMedium.length > 0) {
    chart += '    line [' + rpsMedium.map(d => d.value).join(', ') + ']\n';
  }
  
  if (rpsHigh.length > 0) {
    chart += '    line [' + rpsHigh.map(d => d.value).join(', ') + ']\n';
  }
  
  chart += '```\n';
  
  return chart;
}

/**
 * Generate Mermaid xychart for latency trends
 * @param {Array} historicalData - Historical performance data
 * @returns {string} Mermaid xychart code
 */
function generateLatencyChart(historicalData) {
  const latencyLow = extractScenarioData(historicalData, 'low_concurrent', 'latency_avg');
  const latencyMedium = extractScenarioData(historicalData, 'medium_concurrent', 'latency_avg');
  const latencyHigh = extractScenarioData(historicalData, 'high_concurrent', 'latency_avg');
  
  // Find min and max values for y-axis
  const allValues = [
    ...latencyLow.map(d => d.value),
    ...latencyMedium.map(d => d.value),
    ...latencyHigh.map(d => d.value)
  ];
  const minValue = Math.min(...allValues) * 0.9;
  const maxValue = Math.max(...allValues) * 1.1;
  
  let chart = '```mermaid\nxychart-beta\n    title "Average Latency Trends"\n';
  chart += `    x-axis [${latencyLow.map(d => d.date).join(', ')}]\n`;
  chart += `    y-axis "${minValue}" --> "${maxValue}"\n`;
  chart += '    line [';
  
  if (latencyLow.length > 0) {
    chart += latencyLow.map(d => d.value).join(', ');
  } else {
    chart += '0';
  }
  chart += ']\n';
  
  if (latencyMedium.length > 0) {
    chart += '    line [' + latencyMedium.map(d => d.value).join(', ') + ']\n';
  }
  
  if (latencyHigh.length > 0) {
    chart += '    line [' + latencyHigh.map(d => d.value).join(', ') + ']\n';
  }
  
  chart += '```\n';
  
  return chart;
}

/**
 * Generate Mermaid xychart for P99 latency trends
 * @param {Array} historicalData - Historical performance data
 * @returns {string} Mermaid xychart code
 */
function generateP99LatencyChart(historicalData) {
  const p99Low = extractScenarioData(historicalData, 'low_concurrent', 'latency_p99');
  const p99Medium = extractScenarioData(historicalData, 'medium_concurrent', 'latency_p99');
  const p99High = extractScenarioData(historicalData, 'high_concurrent', 'latency_p99');
  
  // Find min and max values for y-axis
  const allValues = [
    ...p99Low.map(d => d.value),
    ...p99Medium.map(d => d.value),
    ...p99High.map(d => d.value)
  ];
  const minValue = Math.min(...allValues) * 0.9;
  const maxValue = Math.max(...allValues) * 1.1;
  
  let chart = '```mermaid\nxychart-beta\n    title "P99 Latency Trends"\n';
  chart += `    x-axis [${p99Low.map(d => d.date).join(', ')}]\n`;
  chart += `    y-axis "${minValue}" --> "${maxValue}"\n`;
  chart += '    line [';
  
  if (p99Low.length > 0) {
    chart += p99Low.map(d => d.value).join(', ');
  } else {
    chart += '0';
  }
  chart += ']\n';
  
  if (p99Medium.length > 0) {
    chart += '    line [' + p99Medium.map(d => d.value).join(', ') + ']\n';
  }
  
  if (p99High.length > 0) {
    chart += '    line [' + p99High.map(d => d.value).join(', ') + ']\n';
  }
  
  chart += '```\n';
  
  return chart;
}

/**
 * Generate Mermaid xychart for test scenarios trend
 * @param {Array} historicalData - Historical performance data
 * @returns {string} Mermaid xychart code
 */
function generateScenariosChart(historicalData) {
  const dates = historicalData.map(d => formatDate(d.timestamp || d.date || d.updated_at));
  const scenarioCounts = historicalData.map(d => (d.test_scenarios || []).length);
  
  const minValue = Math.min(...scenarioCounts) * 0.9;
  const maxValue = Math.max(...scenarioCounts) * 1.1;
  
  let chart = '```mermaid\nxychart-beta\n    title "Test Scenarios Trends"\n';
  chart += `    x-axis [${dates.join(', ')}]\n`;
  chart += `    y-axis "${minValue}" --> "${maxValue}"\n`;
  chart += '    line [' + scenarioCounts.join(', ') + ']\n';
  chart += '```\n';
  
  return chart;
}

/**
 * Generate complete trend chart report
 * @param {Array} historicalData - Historical performance data
 * @returns {string} Complete Markdown report with Mermaid charts
 */
function generateTrendReport(historicalData) {
  let report = '# UVHTTP Performance Trends\n\n';
  report += '##  Performance Trend Analysis\n\n';
  report += 'This report shows the performance trends over time using Mermaid charts.\n\n';
  
  // Add charts
  report += '### Throughput Trends\n\n';
  report += generateThroughputChart(historicalData);
  report += '\n';
  
  report += '### Average Latency Trends\n\n';
  report += generateLatencyChart(historicalData);
  report += '\n';
  
  report += '### P99 Latency Trends\n\n';
  report += generateP99LatencyChart(historicalData);
  report += '\n';
  
  report += '### Test Scenarios Trends\n\n';
  report += generateScenariosChart(historicalData);
  report += '\n';
  
  // Add legend
  report += '## Legend\n\n';
  report += '- **Low Concurrent**: 10 concurrent connections\n';
  report += '- **Medium Concurrent**: 50 concurrent connections\n';
  report += '- **High Concurrent**: 100 concurrent connections\n';
  
  return report;
}

/**
 * Main function
 */
async function main() {
  const args = process.argv.slice(2);
  
  // Parse arguments
  const options = {
    dataFiles: [],
    outputFile: null
  };
  
  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--output' || args[i] === '-o') {
      options.outputFile = args[++i];
    } else {
      options.dataFiles.push(args[i]);
    }
  }
  
  // Validate arguments
  if (options.dataFiles.length === 0) {
    console.error('Error: No data files provided');
    console.error('');
    console.error('Usage: node generate_trend_chart.js <data_files> --output <output_file>');
    console.error('');
    console.error('Options:');
    console.error('  --output, -o <path>    Output file path');
    console.error('');
    console.error('Example:');
    console.error('  node generate_trend_chart.js data1.json data2.json data3.json --output trends.md');
    process.exit(1);
  }
  
  if (!options.outputFile) {
    console.error('Error: --output is required');
    process.exit(1);
  }
  
  // Load historical data
  const historicalData = [];
  for (const dataFile of options.dataFiles) {
    const data = await loadJsonFile(dataFile);
    if (data) {
      historicalData.push(data);
    }
  }
  
  if (historicalData.length === 0) {
    console.error('Error: No valid data files found');
    process.exit(1);
  }
  
  // Sort by date
  historicalData.sort((a, b) => {
    const dateA = new Date(a.timestamp || a.date || a.updated_at);
    const dateB = new Date(b.timestamp || b.date || b.updated_at);
    return dateA - dateB;
  });
  
  console.log(`Loaded ${historicalData.length} data files`);
  
  // Generate trend report
  const report = generateTrendReport(historicalData);
  
  // Write to file
  await fs.writeFile(options.outputFile, report, 'utf-8');
  console.log(` Performance trend chart saved to ${options.outputFile}`);
}

main().catch(error => {
  console.error('Error:', error.message);
  process.exit(1);
});